language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/diagnostic/DiagnosticsHelper.java

```diff
@@ -187,13 +189,13 @@ public Map<String, List<Diagnostic>> getLatestDiagnostics(DocumentServiceContext
         // NOTE: We are not using `project.sourceRoot()` since it provides the single file project uses a temp path and
         // IDE requires the original path.
         Path projectRoot = workspace.projectRoot(context.filePath());
-        if (project.get().kind() == ProjectKind.SINGLE_FILE_PROJECT) {
-            projectRoot = projectRoot.getParent();
-        }
-        PackageCompilation compilation = workspace.waitAndGetPackageCompilation(context.filePath()).orElseThrow();
+        Path originalPath = project.get().kind() == ProjectKind.SINGLE_FILE_PROJECT
+                ? projectRoot.getParent() : projectRoot;
+        Optional<PackageCompilation> compilationResult = workspace.waitAndGetPackageCompilation(context.filePath(),
+                context.operation() == LSContextOperation.TXT_DID_CHANGE);
```

Comment:
Shall we add `DID_OPEN` as well to here?

Reply:
I don't get the requirement though. I think it's not needed since we check this boolean value, iff the previous compilation has crashed (https://github.com/ballerina-platform/ballerina-lang/pull/39229/files#diff-6d4a15825d022264795ed5bed5a6474047b13e15b8329dbc9fce4d5cfa5ca5ebR305).

Assuming DID_OPEN happens for the below scenarios
1. Opening it on a new tab
2. Move back to the tab

The project can be single or multi-filed.
If it's a single file; the previous compilation crashed means that the file has already been opened. Moving back to the tab does not actually need a compilation because nothing changed.
If it's a multi-filed project, let's say the project crashed due to an issue in a separate file and we are opening another file of the project; even in that case, there's no advantage of recompiling. 
If it's a multi-filed project and we are opening the culprit file itself, even then, without a source change being done, there is no need for recompilation.

wdyt?

Reply:
Yes, agreed. There can be an edge case with generated sources though. We can neglect that at this stage.